src/common/njord.ts

```diff
@@ -452,276 +481,446 @@ export const awardUser = async (
-        entityManager,
-        productId: product.id,
-        receiverId: post.authorId,
-        note,
-      });
+        const transaction = await createTransaction({
+          ctx,
+          id: newTransactionId,
+          entityManager,
+          productId: product.id,
+          receiverId: post.authorId,
+          note,
+        });
 
-      if (!transaction.productId) {
-        throw new Error('Product missing from transaction');
-      }
+        if (!transaction.productId) {
+          throw new Error('Product missing from transaction');
+        }
 
-      await entityManager
-        .getRepository(UserPost)
-        .createQueryBuilder()
-        .insert()
-        .into(UserPost)
-        .values({
-          postId: post.id,
-          userId: ctx.userId,
-          awardTransactionId: transaction.id,
-          flags: {
-            awardId: transaction.productId,
-          },
-        })
-        .onConflict(
-          `("postId", "userId") DO UPDATE SET "awardTransactionId" = EXCLUDED."awardTransactionId", "flags" = user_post.flags || EXCLUDED."flags"`,
-        )
-        .execute();
-
-      const transfer = await transferCores({
-        ctx,
-        transaction,
-      });
+        await entityManager
+          .getRepository(UserPost)
+          .createQueryBuilder()
+          .insert()
+          .into(UserPost)
+          .values({
+            postId: post.id,
+            userId: ctx.userId,
+            awardTransactionId: transaction.id,
+            flags: {
+              awardId: transaction.productId,
+            },
+          })
+          .onConflict(
+            `("postId", "userId") DO UPDATE SET "awardTransactionId" = EXCLUDED."awardTransactionId", "flags" = user_post.flags || EXCLUDED."flags"`,
+          )
+          .execute();
+
+        try {
+          const transfer = await transferCores({
+            ctx,
+            transaction,
+          });
 
-      return { transaction, transfer };
-    },
-  );
+          return { transaction, transfer };
+        } catch (error) {
+          if (error instanceof TransferError) {
+            await entityManager.getRepository(UserPost).delete({
+              awardTransactionId: transaction.id,
+            });
+
+            await throwUserTransactionError({
+              entityManager,
+              error,
+              transaction,
+            });
+          }
 
-  return {
-    transactionId: transaction.id,
-    balance: {
-      amount: parseBigInt(transfer.senderBalance!.newBalance),
-    },
-  };
+          throw error;
+        }
+      },
+    );
+
+    return {
+      transactionId: transaction.id,
+      balance: {
+        amount: parseBigInt(transfer.senderBalance!.newBalance),
+      },
+    };
+  } catch (error) {
+    logger.error(
+      {
+        context: 'award',
+        err: error,
+        props,
+        userId: ctx.userId,
+        transactionId: newTransactionId,
+      },
+      'Award error',
+    );
```

Comment:
In case of critical error we log all the data for mutation, current user and transaction id. That way even if transaction is not commited on API side after njord is finished we can still deduce from which njord transfer error occured based on `transactionId` and `idempotencyKey` in njord.